haskell/test/Test/Util.hs

```diff
@@ -91,14 +96,53 @@ evalJust = \case
   Nothing -> failure "Expected 'Just', got 'Nothing'"
   Just a -> pure a
 
--- | Runs a test for each possible token type combination.
+-- | Runs a test, for each possible token type combination, on the emulator.
 forAllTokenTypeCombinations :: TestName -> ((TokenType, TokenType) -> TestTree) -> TestTree
 forAllTokenTypeCombinations testName mkTest =
   testGroup testName $ do
     x <- xTokenTypes
     y <- yTokenTypes
     pure $ mkTest (x, y)
 
+-- | Runs a test, for each possible token type combination, on the emulator.
+-- For the (FA, CTEZ) combination, the test will also be run on a network.
+forAllTokenTypeCombinationsOnNetwork
+  :: TestName
+  -> (forall m. Monad m => (TokenType, TokenType) -> NettestT m ())
+  -> TestTree
+forAllTokenTypeCombinationsOnNetwork testName mkTest =
+  testGroup testName do
+    x <- xTokenTypes
+    y <- yTokenTypes
+    let tokenTypes = (x, y)
+    if tokenTypes == defaultTokenTypes
+      then one $ nettestScenarioCaps (show tokenTypes) $ mkTest tokenTypes
+      else one $ nettestScenarioOnEmulator (show tokenTypes) \_ -> uncapsNettest (mkTest tokenTypes)
+
+-- | Runs a property test:
+--   * 100 times on the emulator with arbitrary data
+--   * Once on a network with fixed data
+propOnNetwork
```

Comment:
Nice approach, one question: what do you think about generating the data for the network test too? 
I mean basically generate one scenario every time to run on the network.

I suppose in many cases it wouldn't be applicable, but maybe it would work in some.

Also, if so, I wonder if we should add this function to `cleveland` itself.

Reply:
> what do you think about generating the data for the network test too?
I mean basically generate one scenario every time to run on the network.

The problem is that, to embed a Cleveland scenario in hedgehog, you have to use cleveland's `nettestTestProp` , and `nettestTestProp` only runs the test on an emulator.

If we wanted to run it on a network, then we can't embed it in hedgehog.

Technically, we could run a hedgehog generator in IO using [`Gen.sample`](https://hackage.haskell.org/package/hedgehog-1.0.5/docs/Hedgehog-Gen.html#v:sample), and then feed the sample into a Cleveland network test (without actually using properties, see pseudocode below), but then tests wouldn't be reproducible. I.e., you couldn't reproduce a failing network test using `--hedgehog-replay`.

```hs
hprop_myprop :: IO TestTree
hprop_myprop = do
  data <- Gen.sample
  pure $ nettestScenarioOnNetworkCaps "" do
    <use data>
```

Reply:
> we could run a hedgehog generator in IO

yeah that's what I had in mind

> but then tests wouldn't be reproducible

ah right, that's a problem :confused: never mind then :ok_hand: